common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+                        promise completeWith future
+                    }
+            }
+
+        } else {
+            // not caching
+            promise completeWith future
+        }
+
+        promise
+    }
+
+    def getSize() = cache.size
+
+    /**
+     * Log a cache hit
+     *
+     */
+    private def makeNoteOfCacheHit(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_HIT, s"[GET] serving from cache: $key")(logger)
+    }
+
+    /**
+     * Log a cache miss
+     *
+     */
+    private def makeNoteOfCacheMiss(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_MISS, s"[GET] serving from datastore: $key")(logger)
+    }
+
+    /**
+     * We have initiated a read (in cacheLookup), now handle its completion
+     *
+     */
+    private def listenForReadDone(key: Any, entry: Entry, future: => Future[W], promise: Promise[W])(
+        implicit logger: Logging): Unit = {
+
+        future onComplete {
+            case Success(value) => entry.readDone(value, promise)
+            case Failure(t)     => readOops(key, entry, promise, t)
+        }
+    }
+
+    /**
+     * We have initiated a write, now handle its completion
+     *
+     */
+    private def listenForWriteDone(doc: W, key: Any, entry: Entry, future: => Future[Winfo], valPromise: Promise[W], promise: Promise[Winfo])(
+        implicit logger: Logging): Unit = {
+
+        future onComplete {
+            case Success(docinfo) => {
+                //
+                // if the datastore write was successful, then transition to the Cached state
+                //
+                logger.debug(this, "Write backend part done, now marking cache entry as done")
+
+                if (!entry.writeDone()) {
+                    logger.error(this, "Concurrent update detected")
+                    writeOops(key, entry, valPromise, promise, new ConcurrentUpdateException)
+
+                } else {
+                    logger.debug(this, s"Write all done $key ${entry.transid} ${entry.state.get} ${getSize()}")
+                    promise success docinfo
+                    valPromise success doc
+                }
+            }
+            case Failure(t) => {
+                //
+                // oops, the datastore write failed. invalidate the cache entry
+                //
+                writeOops(key, entry, valPromise, promise, t)
+            }
+        }
+    }
+
+    /**
+     * Write completion resulted in a failure
+     *
+     */
+    private def readOops(key: Any, entry: Entry, promise: Promise[W], t: Throwable) {
+        entry.invalidate
+        cache remove key
+        promise failure t
+    }
+
+    /**
+     * Write completion resulted in a failure
+     *
+     */
+    private def writeOops(key: Any, entry: Entry, valPromise: Promise[W], promise: Promise[Winfo], t: Throwable) {
```

Comment:
When you do the `Future` transformation outlined above, this method becomes unnecessary and can be fold into `cacheUpdate` to further localize what actually happens.